backend/ftp/file.go

```diff
@@ -0,0 +1,421 @@
+func (f *File) Size() (uint64, error) {
+
+	entry, err := f.stat(context.TODO())
+	if err != nil {
+		return 0, err
+	}
+	return entry.Size, nil
+}
+
+// Location returns a vfs.Location at the location of the file. IE: if file is at
+// ftp://someuser@host.com/here/is/the/file.txt the location points to ftp://someuser@host.com/here/is/the/
+func (f *File) Location() vfs.Location {
+	return &Location{
+		fileSystem: f.fileSystem,
+		path:       path.Dir(f.path),
+		Authority:  f.authority,
+	}
+}
+
+// Move/Copy Operations
+
+// MoveToFile puts the contents of File into the targetFile passed using File.CopyToFile.
+// If the copy succeeds, the source file is deleted. Any errors from the copy or delete are
+// returned.
+// If the given location is also ftp AND for the same user and host, the ftp Rename method is used, otherwise
+// we'll do a an io.Copy to the destination file then delete source file.
+func (f *File) MoveToFile(t vfs.File) error {
+	// ftp rename if vfs is ftp and for the same user/host
+	if f.fileSystem.Scheme() == t.Location().FileSystem().Scheme() &&
+		f.authority.UserInfo().Username() == t.(*File).authority.UserInfo().Username() &&
+		f.authority.Host() == t.(*File).authority.Host() {
+
+		// ensure destination exists before moving
+		exists, err := t.Location().Exists()
+		if err != nil {
+			return err
+		}
+		dc, err := dataConnGetterFunc(context.TODO(), f, types.SingleOp)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			// it doesn't matter which client we use since they are effectively the same
+			err = dc.MakeDir(t.Location().Path())
+			if err != nil {
+				return err
+			}
+		}
+		return dc.Rename(f.Path(), t.Path())
+	}
+
+	// otherwise do copy-delete
+	if err := f.CopyToFile(t); err != nil {
+		return err
+	}
+	return f.Delete()
+}
+
+// MoveToLocation works by creating a new file on the target location then calling MoveToFile() on it.
+func (f *File) MoveToLocation(location vfs.Location) (vfs.File, error) {
+	newFile, err := location.NewFile(f.Name())
+	if err != nil {
+		return nil, err
+	}
+
+	err = f.MoveToFile(newFile)
+	if err != nil {
+		return nil, err
+	}
+	return newFile, nil
+}
+
+// CopyToFile puts the contents of File into the targetFile passed.
+func (f *File) CopyToFile(file vfs.File) error {
+	if err := backend.ValidateCopySeekPosition(f); err != nil {
+		return err
+	}
+
+	if f.fileSystem.Scheme() == file.Location().FileSystem().Scheme() &&
+		f.authority.UserInfo().Username() == file.(*File).authority.UserInfo().Username() &&
+		f.authority.Host() == file.(*File).authority.Host() {
+		// in the case that both files have the same authority we'll copy by writing a temporary
+		// file to mem and then writing it back to the ftp server
+		tempFile, err := f.createLocalTempFile()
+		if err != nil {
+			return err
+		}
+		if err := utils.TouchCopyBuffered(tempFile, f, 0); err != nil {
+			return err
+		}
+		if err := f.Close(); err != nil {
+			return err
+		}
```

Comment:
We can't close it yet since we still have to read from it in the next TouchCopyBuffered call. Good call though - added logic to seek to start of file before copying from the file https://github.com/C2FO/vfs/pull/141/commits/c059dbf124b3e50dcccaf820e04b5b1967d01f86